### ISSUE TITLE
GPU MNIST fastpath returns CPU-identical logits — silent no-op vs FP16 quantization-noise indistinguishable

### ISSUE THREAD

**Issue by johnjezl** [SLM-OS/SLM-Operating-System#692]
## Symptom

`gpu use inference on` + `lua-admin /mnt/files/mnist_loop.lua` produces logits **bit-identical to 7 decimal places** vs `gpu use inference off` (CPU NEON). Verified on jetson-nano-2 after kexec from Linux with helper-staged channel handoff.

For d0.bin with `gpu use inference on` (HMMA tier helper-staged, --qmd-pool):
```
logit[0] = -0.8843725
logit[1] = 5.1902127
...
```
For d0.bin with `gpu use inference off` (CPU NEON):
```
logit[0] = -0.8843725
logit[1] = 5.1902127
...
```

A simulation of the FC layer with FP16 weights vs FP32 weights on plausible activations produces ~1e-3 deviation at the 4th decimal place — so HMMA tier (FP16 W_fc × FP32 act) **must** produce visibly different logits if it's actually running.

## What's been verified to work

- Fastpath eligibility check passes (`mnist_gpu_fastpath_eligible == true`)
- No `[engine] mnist GPU fastpath failed -- falling back to CPU` log lines on serial across 32+ inferences
- `slm_gpu_run_mnist` returns >= 0 (success)
- `ga10b_dispatch_v7_pipeline` reaches `payload_matched` branch (semaphore release observed)
- GPU consumer toggle reports `inference: ON`
- v6 → v7 handoff wired correctly (helper writes pipeline_n_ops=8, qmd_pool_n_slots=1024)
- HMMA shader bytes staged at `gemm_shader_gva` (helper logs `tier=hmma`, gemm=7552 bytes)
- v7 op fields propagate `smem_size_bytes=2048` and `barrier_count=3` for op[6] (PR adding the propagation closes a structural gap; see PR description)

## Hypotheses

1. **MNIST FC math is robust to FP16 weight quantization at FP32 logit precision for this specific (model × inputs) pair.** Plausible only if the trained weights happen to be FP16-representable to ULP — unlikely from random simulation but possible for a quantization-aware-trained model.
2. **GPU dispatch is silently completing without actually executing the SM** (e.g. SKED accepts the QMD, advances GP_PUT, releases the semaphore, but the kernel never runs because `BARRIER_COUNT=3` interferes with WMMA in some way the helper's Linux self-test masks). Cells in op[6].output buffer would stay zeroed; op[7] AddBias would compute on zeros and produce a known constant → matches identical-to-CPU only if CPU also somehow converges. Doesn't fit.
3. **Some path silently routes to CPU even when gpu inference is ON.** Would need to reproduce by tracing engine.rs's run_inference for evidence — `Ok(n) => return Ok(n)` from fastpath should put GPU bytes in the output, but if `read_pipeline_output` reads stale cached zero (CPU cache holds the helper's pre-zeroed lines because GPU writes don't snoop) and the output_phys is in cacheable DRAM, the Rust caller sees zeros — but that contradicts the non-zero logits returned.

Most likely (3) given the bit-equality evidence, but the mechanism is unclear.

## Diagnostic next steps

- Add a one-shot uart_printf in `ga10b_dispatch_v7_pipeline` after the semaphore poll showing the first 4 bytes of `last->output_phys` (post-cache_invalidate) — confirms whether the buffer is non-zero at read time.
- Add a one-shot uart_printf in `read_pipeline_output` showing the same — discriminates cache invalidation issues.
- Peek `0x13ac6d000` from shell **immediately** after a fresh `lua-admin one_infer.lua` (cache-cold from peek's perspective) — if zero, GPU isn't writing there.
- Add a SM activity counter check (NV_PGRAPH_PRI_GPCS_TPCS_SMS_DBGR_BPT_PAUSE_MASK or similar) to confirm SM busy time during dispatch.

## Repro

```
# On jetson-nano-2 Linux:
cd /root/gpu-mnist
./gpu-kernel-mnist --preserve-for-kexec --qmd-pool --gemm-tier hmma \
   --weights-fp16-dir mnist-weights-fp16 --weights-dir mnist-weights \
   --shader-dir . --timeout-secs 1800 &
slmos-kexec
# Under SLM-OS:
gpu use inference on
lua-admin /mnt/files/all_logits.lua  # prints 10 logits at 7 decimals
gpu use inference off
lua-admin /mnt/files/all_logits.lua  # compare
```

## Scope

Doesn't block: CPU MNIST under SLM-OS works (deterministic, multi-pass); GPU MNIST appears to work (no failure logs, deterministic across passes); upload + telnet + lua-admin path validated.

Does block: claims that "tensor cores are demonstrably executing under SLM-OS" — the visible numerical signature can't distinguish HMMA from SIMT-FP32 from CPU-NEON-FP32 today.

**Comment by johnjezl** [comment]
## Investigation update (2026-05-07)

Instrumented the dispatch path and reproduced the bit-equality observation. Key findings:

### Boot-log smoking gun

```
[GSP] firmware missing: gsp
[GSP] firmware missing: bootloader
[GSP] firmware missing: booter_load
[GSP] firmware missing: booter_unload
[WARN] GSP init incomplete (rc=-1) — GPU compute unavailable
```

`[WARN] GSP init incomplete (rc=-1) — GPU compute unavailable` fires every kexec boot on jetson-nano-2. (Note: Jetson's GA10B uses `ga10b_bringup_inherit` rather than GSP-RM, so this warning may be benign for the MNIST path — but the warning suggests the C-side `compute_ready=1` flag is decoupled from actual compute readiness.)

### Eligibility / dispatch diagnostic results

Added three diagnostics:
- `[RUST-INFER-DIAG-692] rust_infer entry` at the top of `rust_infer` (lib.rs)
- `[ELIG-DIAG-692] <reason>` in every branch of `mnist_gpu_fastpath_eligible` (engine.rs)
- `[GPU-DIAG-692] slm_gpu_run kind=N rc=R` at entry+exit of `slm_gpu_run` (slm_ffi.c)

After deploying the diagnostic kernel and triggering `lua-admin /mnt/files/all_logits.lua` with `gpu use inference: ON`:

- **None of the three diagnostics fired on serial.**
- No `[engine] mnist GPU fastpath failed -- falling back to CPU` log either.
- Inference completed in normal time and returned the same bit-identical-to-CPU logits.

### Conclusion (provisional)

The Lua MNIST inference path is going through `slm.model_infer_file` → `l_model_infer_file` → `rust_infer`, but on this kernel revision the diagnostic at the top of `rust_infer` doesn't appear on serial. Two possibilities:

1. **The Rust diagnostics aren't reaching UART** for some reason (rate-limit, log-level, task-CPU-affinity, or compiler-elision). C-side `slm_gpu_run` diagnostic also missing — same disposition for non-CPU0 callers, since uart_print on Jetson is IRQ-disable-only and assumes CPU0 print contract (per `kernel/CLAUDE.md` §"UART Lock on Pi 5 / Jetson").

2. **`mnist_gpu_fastpath_eligible` does return false silently somewhere.** With `gpu use inference: ON` confirmed, the most likely failed gate is `caps.has_compute()` — which reads `info.compute_ready` from C, set unconditionally to 1 on Jetson via `#ifdef PLATFORM_JETSON_ORIN_NANO` in `slm_ffi.c::slm_gpu_get_info`. So this gate should pass. Next gate is `name == "mnist"` — model is registered as "mnist". Per-model `gpu_dispatch_enabled` defaults true. Master `slm_gpu_inference_enabled` is on.

The shell `help model` text says: *"GPU dispatch decision: ... when ... a v6 channel handoff is present (Jetson with --no-gpu-suspend kexec)."* The default kexec path uses suspend (drains DMA before kexec), which the help suggests does not produce a usable v6 handoff. Combined with no diagnostic fires, the strongest hypothesis is:

**The default suspend-kexec path from Linux→SLM-OS does not produce a working GA10B channel handoff. `caps.has_compute()` reports true based on a static `#ifdef`, the eligibility check passes, the dispatch path runs, fails internally somewhere — possibly without reaching the `Err(_)` branch in `engine.rs:843` because the failure occurs inside a separate code path (e.g., engine-side CPU fallback that `run_inference` doesn't log) — and the user gets bit-identical-to-CPU output.**

I instrumented Rust-side log calls but they didn't appear on serial. The C-side `slm_gpu_run` diag also didn't appear. Both either (a) genuinely never run, or (b) print but the print is dropped on a non-CPU0 task (the kernel's UART lock contract).

### Recommended next steps

- **Audit the eligibility/dispatch path's logging** to ensure failure modes log unconditionally to UART. Today `engine.rs:843` only logs on `Err(_)` from the fastpath; if `mnist_gpu_fastpath_eligible` returns false (CPU path silently taken), no log fires.
- **Make `compute_ready` actually validate the channel handoff** (not a `#ifdef` alone). The suspended-kexec path doesn't produce a usable handoff, so `compute_ready=1` is currently lying.
- **Add `gpu use status` (or extend it)** to surface `g_handoff.version` and `g_handoff.pipeline_n_ops` so an operator can check from the shell whether GPU compute is actually wired up.
- **Run `--no-gpu-suspend` mode** and see if v6 handoff survives — that's the path the help text endorses for a working GPU fastpath.

I haven't fully resolved the diagnostic-log mystery but the directional evidence is strong: the user-visible "GPU MNIST" path on the default suspend-kexec deployment is silently CPU-only.

The structural fix in #694 (v7 QMD pool propagating smem/barrier overrides) remains correct regardless — it's a precondition for tensor-core dispatch once the channel-handoff issue is fixed.

**Comment by johnjezl** [comment]
Filed #702 to track the architectural fix: make `compute_ready` reflect actual handoff state instead of being a static `#ifdef PLATFORM_JETSON_ORIN_NANO`, and surface handoff diagnostics in `gpu use status`. That's the suspected mechanism behind the bit-identical logits observed here.

**Comment by johnjezl** [comment]
## Resolution: not reproducing on current main (`d28dcde0`)

Drove the experiment from the SLM-OS shell on jetson-nano-2 with `model infer mnist` (which forces the eligibility gate via `slm_gpu_inference_enabled()` — toggle-respecting path, unlike `slm.gpu_run_mnist()` which always dispatches to GPU). Added one-shot diagnostic `uart_printf`s in `ga10b_bringup_read_pipeline_output` and `engine::run_inference` to dump the first 16 bytes at the GPU output_phys + the first 4 logit bytes returned to the Rust caller.

### Four-iteration data

| Run | Path | bytes[0..16] at `last_output_phys` | Time |
|------|------|------------------------------------|--------|
| 1 | CPU (`gpu use inference off`) | `bd3b91fa ba6d8a48 3d912dc1 3d2d2027` | 13.06 ms |
| 2 | GPU (`gpu use inference on`, cold) | `bd37baf4 3bff5131 3d8b7871 3cf5b56f` | **23.50 s** |
| 3 | GPU (hot) | `bd37baf4 3bff5131 3d8b7871 3cf5b56f` | 122 ms |
| 4 | CPU (back to off) | `bd3b91fa ba6d8a48 3d912dc1 3d2d2027` | 13.05 ms |

- **GPU and CPU paths produce distinct deterministic bytes** at every position, not just LSBs.
- The kernel-side `[692-diag] read_pipeline_output` log matches the Rust-side `[692-diag] engine: GPU return` log exactly — the byte path from GPU buffer to Rust output is clean.
- The 23.5 s → 120 ms cold-to-hot transition is unmistakable real-GPU-dispatch signature (`nvgpu inherit` channel warm-up on first call).
- Both paths produce `argmax = 5` for the fixed `model infer` test input. The small numerical differences (`logit[5] = 0.150` CPU vs `0.140` GPU rounded to 3 decimals) don't tip the winner but are clearly distinguishable.

### Likely cause of the original symptom

The original repro used `lua-admin /mnt/files/all_logits.lua`. The Lua binding `slm.gpu_run_mnist()` (`kernel/src/lua_slm.c:699`) calls `slm_gpu_run_mnist()` directly, which **bypasses the eligibility gate entirely** — every invocation goes through the GPU regardless of `gpu use inference` state. So `all_logits.lua`'s two consecutive calls (with toggle on, then toggle off) both went through the GPU and produced byte-identical output. The bit-equality wasn't "GPU silently fell back to CPU" — it was "Lua never asked engine.rs which path to take."

The toggle-respecting path through `slm.model_infer*` / `slm.model_infer_file` / `model infer` shell command produces the visibly-different results above on the same build.

### Hypothesis status

- (1) **MNIST FC math robust to FP16 quantization** — irrelevant; not what was measured. The GPU IS producing different bytes.
- (2) **GPU dispatches but kernel never runs** — disproven; non-zero, non-CPU-equal bytes produced.
- (3) **Silent CPU fallback when GPU is on** — disproven via the toggle-respecting path. The original symptom is explained by the `slm.gpu_run_mnist()` binding's intentional toggle-bypass, not by a silent fallback.

### Closing

No code fix needed. Recommend that future regression coverage for "is the GPU actually executing" go through `model infer` (shell) or `slm.model_infer*` (Lua bindings that respect the toggle) rather than `slm.gpu_run_mnist()` directly. If `all_logits.lua` is used for benchmarking, it should be updated to call `slm.model_infer(idx)` and compare against the CPU-toggle path to give a meaningful GPU-vs-CPU signal.

Reopening welcome if the symptom recurs on the toggle-respecting path. The diagnostic instrumentation used for this experiment is preserved in the conversation transcript and can be re-applied as a 2-line patch in `engine.rs` + `ga10b_bringup.c::ga10b_bringup_read_pipeline_output`.

Closed at `d28dcde0`.
